soc/arm/silabs_sim3/sim3u/soc.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2023 GARDENA GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <zephyr/irq.h>
+
+#include <SI32_CLKCTRL_A_Type.h>
+#include <SI32_PBCFG_A_Type.h>
+#include <SI32_PBSTD_A_Type.h>
+#include <SI32_RSTSRC_A_Type.h>
+#include <SI32_VMON_A_Type.h>
+#include <SI32_WDTIMER_A_Type.h>
+#include <si32_device.h>
+#include <silabs_crossbar_config.h>
+
+static void gpio_init(void)
+{
+	/* Enable clocks that we need in any case */
+	SI32_CLKCTRL_A_enable_apb_to_modules_0(
+		SI32_CLKCTRL_0, SI32_CLKCTRL_A_APBCLKG0_PB0 | SI32_CLKCTRL_A_APBCLKG0_FLASHCTRL0);
+
+	/* Enable misc clocks.
+	 * We may or may nor need all of that but it includes some basics like
+	 * LOCK0, WDTIMER0 and DMA. Doing it here prevents the actual drivers
+	 * needing to know whoelse needs one of these clocks.
+	 */
+	SI32_CLKCTRL_A_enable_apb_to_modules_1(
+		SI32_CLKCTRL_0, SI32_CLKCTRL_A_APBCLKG1_MISC0 | SI32_CLKCTRL_A_APBCLKG1_MISC1);
+	SI32_CLKCTRL_A_enable_ahb_to_dma_controller(SI32_CLKCTRL_0);
+
+	/* Apply pinctrl gpio settings */
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_0, PORTBANK_0_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_0, PORTBANK_0_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_0, PORTBANK_0_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_0, PORTBANK_0_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_0, PORTBANK_0_PINS_ANALOG);
+
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_1, PORTBANK_1_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_1, PORTBANK_1_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_1, PORTBANK_1_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_1, PORTBANK_1_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_1, PORTBANK_1_PINS_ANALOG);
+
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_2, PORTBANK_2_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_2, PORTBANK_2_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_2, PORTBANK_2_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_2, PORTBANK_2_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_2, PORTBANK_2_PINS_ANALOG);
+
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_3, PORTBANK_3_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_3, PORTBANK_3_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_3, PORTBANK_3_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_3, PORTBANK_3_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_3, PORTBANK_3_PINS_ANALOG);
```

Comment:
Shouldn't this be handled by the pinctrl driver then?

Reply:
We don't have a pinctrl driver. That's the reason why this exists in first place. The GPIO driver shouldn't be used here either since it must be set even if `CONFIG_GPIO` is disabled.

Reply:
why is this code put here and not in a pinctrl driver? can you please explain why can't this be executed by each driver?

Reply:
A device driver usually informs the pinctrl subsystem about the pinconfig it wants by calling `pinctrl_apply_state`. After that function returns it expect things to be muxed already and may start using the device.

So what we need to do on sim3u is to know which pinmuxings are needed before even trying to probe every driver.

Of course, we could just cache the request in "pinctrl_apply_state" but not actually apply any changes, then register a SYS_INIT at a very late level like `APPLICATION`, but IMO that would violate the pinctrl API and may not work with certain drivers.

IMO the pinctrl API is just fundamentally incompatible with the way sim3u works and I was simply trying to find a way to at least be able to use the expressiveness provided by the devicetree nodes.